### PR TITLE
Document collection subscriptions

### PR DESCRIPTION
**Pull request by hannako**
Navigation and Presentation will soon be rolling out email subscriptions for document collections. 

The required behaviour for these subscriptions is: When a user subscribes to notifications on a document collection, they will be notified when the document collection is updated AND when any of the documents belonging to that collection are updated. 

The relationship between a document collection and the documents it contains is stored as a reverse link. 

This PR adds support for reverse link style lists in the future ([See here for a list of reverse link types](https://github.com/alphagov/publishing-api/blob/main/lib/expansion_rules.rb#L43-L56)), but currently only allows these new lists to be created for document collections.

Related work: https://github.com/alphagov/email-alert-service/pull/583
Trello: https://trello.com/c/tJvdhfdd/1455-support-topic-like-subscriptions-for-document-collections-email-alert-api-m, [Jira issue NAV-5609](https://gov-uk.atlassian.net/browse/NAV-5609)

⚠️  This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

**Review by chao-xian — APPROVED** [review]
LGTM but I'd like for Kevin and Keith to look too.

**Review by KludgeKML — APPROVED** [review]
LGTM

File: app/models/subscriber_list_params/single_page_lists.rb [hunk 1]
```diff
@@ -16,7 +18,33 @@ def list_params
...
+    def links
+      { "links" => strip_if_empty_key(reverse_links) }
+    end
+
+    def reverse_links
+      { reverse_linked_key => [content_item["content_id"]] }
+    end
+
+    def reverse_linked_key
+      DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS[document_type] || ""
```
Comment (kevindew) [inline app/models/subscriber_list_params/single_page_lists.rb:35]:
```suggestion
      DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS[document_type]
```

I imagine it doesn't matter if this returns nil?

**Review by kevindew — COMMENTED** [review]
Looks good to me 👍 I've added a few implementation suggestions/problems

File: app/controllers/single_page_subscriptions_controller.rb [hunk 2]
```diff
@@ -53,15 +53,19 @@ def topic_id
     @topic_id = params[:topic_id]
   end
 
-  def validate_base_path
-    URI.parse(params.fetch(:base_path))
+  def assign_content_item
+    content_item_path = params.fetch(:base_path)
+
+    return bad_request unless content_item_path.to_s.starts_with?("/")
+    return bad_request unless URI.parse(content_item_path).relative?
```
[the remark this answers is not in the document]
Comment (hannako) [reply app/controllers/single_page_subscriptions_controller.rb:60]: This commit isn't necessary for this PR. I was trying to slowly bring the two controllers into alignment, with a view to perhaps refactoring down the line as there is a lot of shared functionality. So this code change was a straight cut and paste. I think for the sake of clarity I will drop this commit, and do any refactoring as a stand alone piece of work. With new specs etc as needed.

File: app/models/subscriber_list_params/single_page_lists.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 module SubscriberListParams
   class SinglePageLists
+    DOCUMENT_TYPES_PERMITTED_TO_HAVE_REVERSE_LINKED_LISTS = { "document_collection" => "document_collections" }.freeze
```
[the remark this answers is not in the document]
Comment (hannako) [reply app/models/subscriber_list_params/single_page_lists.rb:3]: Yes that's better, thank you

File: app/models/subscriber_list_params/single_page_lists.rb (hunk 1 above)
Comment (hannako) [reply app/models/subscriber_list_params/single_page_lists.rb:35]: Ah yes you're right, thank you

File: app/services/subscriber_list_params/generate_single_page_list_params_service.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      @content_item = content_item
+    end
+
+    def call
+      {
+        "url" => content_item["base_path"],
+        "title" => content_item["title"],
+        "content_id" => content_item["content_id"],
+        "description" => content_item["description"],
+      }.merge(strip_if_empty_value(links))
```
Comment (leenagupte) [inline app/services/subscriber_list_params/generate_single_page_list_params_service.rb:16]: I'm unsure of the need for the `strip_if_empty_value` and `strip_if_empty_key` methods.

I think these checks can probably be done in the `links` method. 

The check here can be replaced with an `present?` check, if the hash is assigned to a variable first.

**Review by leenagupte — COMMENTED** [review]
One comment /  question about the "strip" methods.

The rest is looking good. I especially like how the renaming of files and adding of namespaces has been split into separate commits. That made this much easier to review. ⭐

File: app/services/subscriber_list_params/generate_links_based_list_params_service.rb [hunk 5]
```diff
@@ -1,51 +1,53 @@
...
+    end
 
-private
+  private
 
-  attr_reader :content_item
+    attr_reader :content_item
 
-  class UnsupportedContentItemError < StandardError; end
+    class UnsupportedContentItemError < StandardError; end
```
Comment (kevindew) [inline app/services/subscriber_list_params/generate_links_based_list_params_service.rb:20]: Just semantic, but you should move this out of the private part of this class given it's part of the public interface.

**Review by kevindew — COMMENTED** [review]
This looks good to me aside a minor comment. Though I agree with Leena's comment too.

File: app/services/subscriber_list_params/generate_single_page_list_params_service.rb (hunk 4 above)
Comment (hannako) [reply app/services/subscriber_list_params/generate_single_page_list_params_service.rb]: Thanks Leena - not sure why I made that so complicated. Have refactored 👍

File: app/services/subscriber_list_params/generate_links_based_list_params_service.rb (hunk 5 above)
Comment (hannako) [reply app/services/subscriber_list_params/generate_links_based_list_params_service.rb]: Thanks, have made that change.

**Comment by hannako** [comment]
@kevindew I've made the changes requested - in the absence of Leena and Kelv, are you happy for me to merge this?

**Review by kevindew — APPROVED** [review]

**Pull request merged**